### PR TITLE
interfaces/builtin: add storage-framework-service interface

### PR DESCRIPTION
**Pull request by michihenning**
This adds a snapd interface for use by storage-framework. On the slot side, it allows use of aa_is_enabled() and binding to the storage framework registry and provider DBus interfaces. On the plug side, it allows clients to talk to these DBus interfaces.

### CODE REVIEW

File: interfaces/builtin/basedeclaration_test.go [hunk 1]
```diff
@@ -128,11 +128,11 @@ func (s *baseDeclSuite) TestAutoConnection(c *C) {
 	// these have more complex or in flux policies and have their
 	// own separate tests
 	snowflakes := map[string]bool{
-		"content":       true,
-		"core-support":  true,
-		"home":          true,
-		"lxd-support":   true,
-		"snapd-control": true,
+		"content":           true,
```
Comment (zyga) [inline interfaces/builtin/basedeclaration_test.go]: This seems unrelated, did you change it explicitly or did `go ftm` decided to do so?

File: interfaces/builtin/storage_framework.go [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+    bus=session
+    interface=com.canonical.StorageFramework.Provider.*
+    path=/provider/*
+    peer=(label=###SLOT_SECURITY_TAGS###),
+`
+
+type StorageFrameworkInterface struct{}
+
+func (iface *StorageFrameworkInterface) Name() string {
+	return "storage-framework"
```
Comment (zyga) [inline interfaces/builtin/storage_framework.go]: Storage framework is a generic name. Is there any documentation for what the canonical storage framework is doing?

File: interfaces/builtin/basedeclaration_test.go (hunk 1 above)
Comment (michihenning) [inline interfaces/builtin/basedeclaration_test.go]: That wasn't me. I think go fmt is responsible.

File: interfaces/builtin/storage_framework.go (hunk 2 above)
Comment (michihenning) [inline interfaces/builtin/storage_framework.go]: We have an lp project for it: lp:storage-framework. We are waiting for management to allocate time to write documentation.

**Comment by mvo5** [comment]
There is a test failure related to formating: https://travis-ci.org/snapcore/snapd/jobs/204858792#L408 - you may need to run `gofmt -s -w` on these file(s).

**Comment by mvo5** [comment]
I pushed the `go fmt` fix.

**Comment by michihenning** [comment]
> I pushed the go fmt fix.

Thanks for that! My apologies for not replying sooner. Somehow, I missed the notification email for your comments.

**Comment by niemeyer** [comment]
@jdstrand This should be an easy one.

**Comment by michihenning** [comment]
Just checking: the thumbnailer is a service too, and its interface is called "thumbnailer". I chose "storage-framework" for consistency. I don't mind changing it to "storage-framework-service". Just let me know.

File: interfaces/builtin/storage_framework.go [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+	return nil, nil
+}
+
+func (iface *StorageFrameworkInterface) ConnectedSlotSnippet(plug *interfaces.Plug, slot *interfaces.Slot, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	switch securitySystem {
+	case interfaces.SecurityAppArmor:
+		snippet := []byte(storageFrameworkConnectedSlotAppArmor)
+		old := []byte("###PLUG_SNAP_NAME###")
+		new := []byte(plug.Snap.Name())
+		snippet = bytes.Replace(snippet, old, new, -1)
```
Comment (jdstrand) [inline interfaces/builtin/storage_framework.go]: ###PLUG_SNAP_NAME### isn't currently in storageFrameworkConnectedSlotAppArmor. Was it meant to be used as part of the `path` in this rule:
```
dbus (receive, send)
    bus=session
    interface=com.canonical.StorageFramework.Provider.*
    path=/provider/*
    peer=(label=###PLUG_SECURITY_TAGS###),
```

If so, then I like that idea a lot.

File: interfaces/builtin/storage_framework_test.go [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+	c.Assert(snippet, Not(IsNil))
+	// slots have a permanent non-nil security snippet for apparmor
+	snippet, err = s.iface.PermanentSlotSnippet(s.slot, interfaces.SecurityAppArmor)
+	c.Assert(err, IsNil)
+	c.Assert(snippet, Not(IsNil))
+	// plugs have a non-nil security snippet for apparmor
+	snippet, err = s.iface.ConnectedPlugSnippet(s.plug, s.slot, interfaces.SecurityAppArmor)
+	c.Assert(err, IsNil)
+	c.Assert(snippet, Not(IsNil))
+}
```
Comment (jdstrand) [inline interfaces/builtin/storage_framework_test.go]: It would be nice to add a test that the rules you are adding for ConnectedSlot and ConnectedPlug policy are preset similar to thumbnailer_test.go in TestSlotGrantedAccessToPlugFiles().

**Comment by niemeyer** [comment]
@michihenning We can't have interfaces called "online-accounts" or "storage-framework" when they are talking about very specific software implementations. snapd is widely scoped and a multi-Linux-distribution project, so we need to think about how much sense it makes for someone to be siting at a console in a completely different place and looking at these names. "thumbnailer" feels bad as well, actually. I'd prefer to rename that one too.

File: interfaces/builtin/storage_framework.go (hunk 3 above)
Comment (michihenning) [inline interfaces/builtin/storage_framework.go]: The storage framework doesn't need access to any files on the plug side, so PLUG_SNAP_NAME isn't needed here.

Comment (michihenning) [inline interfaces/builtin/storage_framework.go]: The substitution was a copy-and-paste error and was redundant. I've removed it.

File: interfaces/builtin/storage_framework_test.go (hunk 4 above)
Comment (michihenning) [inline interfaces/builtin/storage_framework_test.go]: See above. Because the service doesn't access to anything on the plug side, there is no need to test whether access is granted.

**Comment by michihenning** [comment]
I think I've made all the changes. The autopkg test failures are unrelated to this pull request.

I'll do a separate pull request to rename thumbnailer -> thumbnailer-service.

File: interfaces/builtin/basedeclaration.go [hunk 5]
```diff
@@ -516,6 +516,12 @@ slots:
       slot-snap-type:
         - core
     deny-auto-connection: true
+  storage-framework-service:
+    allow-installation:
+      slot-snap-type:
+        - app
+    deny-connection: true
+    deny-auto-connection: true
```
Comment (jdstrand) [inline interfaces/builtin/basedeclaration.go]: The base declaration LGTM.

File: interfaces/builtin/basedeclaration_test.go [hunk 6]
```diff
@@ -572,20 +573,21 @@ func (s *baseDeclSuite) TestConnection(c *C) {
...
-		"ubuntu-download-manager": true,
+		"bluez":            true,
+		"content":          true,
+		"docker":           true,
+		"fwupd":            true,
+		"location-control": true,
+		"location-observe": true,
+		"lxd":              true,
+		"maliit":           true,
+		"mir":              true,
```
Comment (jdstrand) [inline interfaces/builtin/basedeclaration_test.go]: I think you need to run go fmt again to fix these.

File: interfaces/builtin/storage_framework_service.go [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/interfaces/apparmor"
+)
+
+const storageFrameworkServicePermanentSlotAppArmor = `
+# Description: Allow use of aa_is_enabled()
+
+/sys/module/apparmor/parameters/enabled r,
+@{PROC}/@{pid}/mounts                   r,
+/sys/kernel/security/apparmor/.access   rw,
```
Comment (jdstrand) [inline interfaces/builtin/storage_framework_service.go]: Can you add a comment for this? I suggest:
```
# libapparmor query interface needs 'w' to perform the query and 'r' to
# read the result. This is an information leak because in addition to
# allowing querying policy for any label (precisely what
# storage-framework needs), it also allows checking the existence of
# any label.
/sys/kernel/security/apparmor/.access rw,
```

File: interfaces/builtin/storage_framework_service.go [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+    path=/com/canonical/StorageFramework/Registry
+    peer=(label=###SLOT_SECURITY_TAGS###),
+
+dbus (receive, send)
+    bus=session
+    interface=com.canonical.StorageFramework.Provider.*
+    path=/provider/*
+    peer=(label=###SLOT_SECURITY_TAGS###),
+`
+
```
Comment (jdstrand) [inline interfaces/builtin/storage_framework_service.go]: Can you add storageFrameworkServicePermanentSlotSecComp with the following:
```
const upowerObservePermanentSlotSeccomp = `
bind
`
```

Your storage-framework snap may be specifying 'plugs: [ network-bind ]' that is giving you this syscall, but the slots side should have it on its own to bind to the DBus well-known name.

File: interfaces/builtin/storage_framework_service.go [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+	new := slotAppLabelExpr(slot)
+	snippet = strings.Replace(snippet, old, new, -1)
+	spec.AddSnippet(snippet)
+	return nil
+}
+
+func (iface *StorageFrameworkServiceInterface) AppArmorPermanentSlot(spec *apparmor.Specification, slot *interfaces.Slot) error {
+	spec.AddSnippet(storageFrameworkServicePermanentSlotAppArmor)
+	return nil
+}
```
Comment (jdstrand) [inline interfaces/builtin/storage_framework_service.go]: Can you add:
```
func (iface *StorageFrameworkServiceInterface) SecCompPermanentSlot(spec *seccomp.Specification, slot *interfaces.Slot) error {
	spec.AddSnippet(storageFrameworkServicePermanentSlotSecComp)
	return nil
}
```

File: interfaces/builtin/storage_framework_service.go [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+
+func (iface *StorageFrameworkServiceInterface) PermanentSlotSnippet(slot *interfaces.Slot, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	return nil, nil
+}
+
+func (iface *StorageFrameworkServiceInterface) AppArmorConnectedSlot(spec *apparmor.Specification, plug *interfaces.Plug, slot *interfaces.Slot) error {
+	snippet := storageFrameworkServiceConnectedSlotAppArmor
+	old := "###PLUG_SECURITY_TAGS###"
+	new := slotAppLabelExpr(slot)
```
Comment (jdstrand) [inline interfaces/builtin/storage_framework_service.go]: This should be:
```
new := plugAppLabelExpr(plug)
```

File: interfaces/builtin/storage_framework_service_test.go [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+  command: foo
+  slots: [storage-framework-service]
+`
+	// a storage-framework-service slot on the core snap (as automatically added on classic)
+	const storageFrameworkServiceMockClassicSlotSnapInfoYaml = `name: core
+type: os
+slots:
+ storage-framework-service:
+  interface: storage-framework-service
+`
```
Comment (jdstrand) [inline interfaces/builtin/storage_framework_service_test.go]: You should drop storageFrameworkServiceMockClassicSlotSnapInfoYaml because the storage-framework-service interface is not an implicit interface and the basedeclaration says only snaps of type 'app' may slot this interface.

File: interfaces/builtin/storage_framework_service_test.go [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+  command: foo
+  plugs: [storage-framework-service]
+`
+	s.iface = &builtin.StorageFrameworkServiceInterface{}
+	// storage-framework-service snap with storage-framework-service slot on an core/all-snap install.
+	snapInfo := snaptest.MockInfo(c, storageFrameworkServiceMockCoreSlotSnapInfoYaml, nil)
+	s.coreSlot = &interfaces.Slot{SlotInfo: snapInfo.Slots["storage-framework-service"]}
+	// storage-framework-service slot on a core snap in a classic install.
+	snapInfo = snaptest.MockInfo(c, storageFrameworkServiceMockClassicSlotSnapInfoYaml, nil)
+	s.classicSlot = &interfaces.Slot{SlotInfo: snapInfo.Slots["storage-framework-service"]}
```
Comment (jdstrand) [inline interfaces/builtin/storage_framework_service_test.go]: Please remove these classic checks (see above).

File: interfaces/builtin/storage_framework_service_test.go [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+	c.Check(func() { s.iface.SanitizePlug(&interfaces.Plug{PlugInfo: &snap.PlugInfo{Interface: "other"}}) },
+		PanicMatches, `plug is not of interface "storage-framework-service"`)
+}
+
+func (s *StorageFrameworkServiceInterfaceSuite) TestUsedSecuritySystems(c *C) {
+	// connected slots have a non-nil security snippet for apparmor
+	apparmorSpec := &apparmor.Specification{}
+	err := apparmorSpec.AddConnectedSlot(s.iface, s.plug, s.coreSlot)
+	c.Assert(err, IsNil)
+	c.Assert(apparmorSpec.SecurityTags(), DeepEquals, []string{"snap.storage-framework-service.app"})
```
Comment (jdstrand) [inline interfaces/builtin/storage_framework_service_test.go]: This will need to be changed to "snap.client.app" due to the requested changes in AppArmorConnectedSlot() (see above).

File: interfaces/builtin/storage_framework_service_test.go [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+	err = apparmorSpec.AddConnectedSlot(s.iface, s.plug, s.classicSlot)
+	c.Assert(err, IsNil)
+	c.Assert(apparmorSpec.SecurityTags(), HasLen, 0)
+
+	// slots have a permanent non-nil security snippet for apparmor
+	apparmorSpec = &apparmor.Specification{}
+	err = apparmorSpec.AddPermanentSlot(s.iface, s.coreSlot)
+	c.Assert(err, IsNil)
+	c.Assert(apparmorSpec.SecurityTags(), DeepEquals, []string{"snap.storage-framework-service.app"})
+	c.Assert(apparmorSpec.SnippetForTag("snap.storage-framework-service.app"), testutil.Contains, `member={RequestName,ReleaseName,GetConnectionCredentials}`)
```
Comment (jdstrand) [inline interfaces/builtin/storage_framework_service_test.go]: Can you also check for the seccomp "bind". Eg:
```
seccompSpec := &seccomp.Specification{}
err := seccompSpec.AddPermanentSlot(s.iface, s.coreSlot)
c.Assert(err, IsNil)
c.Assert(seccompSpec.SecurityTags(), DeepEquals, []string{"snap.storage-framework-service.app"})
c.Check(seccompSpec.SnippetForTag("snap.storage-framework-service.app"), testutil.Contains, "bind\n")
```

File: interfaces/builtin/storage_framework_service_test.go [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+	c.Assert(err, IsNil)
+	c.Assert(apparmorSpec.SecurityTags(), DeepEquals, []string{"snap.storage-framework-service.app"})
+	c.Assert(apparmorSpec.SnippetForTag("snap.storage-framework-service.app"), testutil.Contains, `interface=com.canonical.StorageFramework`)
+
+	// slots have no permanent snippet on classic
+	apparmorSpec = &apparmor.Specification{}
+	err = apparmorSpec.AddConnectedSlot(s.iface, s.plug, s.classicSlot)
+	c.Assert(err, IsNil)
+	c.Assert(apparmorSpec.SecurityTags(), HasLen, 0)
+
```
Comment (jdstrand) [inline interfaces/builtin/storage_framework_service_test.go]: Can you add corresponding tests for apparmorSpec.AddConnectedPlug?

File: interfaces/builtin/storage_framework_service_test.go (hunk 14 above)
Comment (michihenning) [inline interfaces/builtin/storage_framework_service_test.go]: Thanks for the review Jamie! I apologize for being so clumsy. Much of this is new to me.
I take it that it is OK not to assign to classicSlot? (This is the one thing I'm not sure about. I think all the other changes you asked for are there now.)

**Comment by zyga** [comment]
Hey @michihenning can you please merge master into this branch?
@niemeyer / @jdstrand can you please re-review this?

File: interfaces/builtin/storage_framework_service_test.go (hunk 14 above)
Comment (jdstrand) [inline interfaces/builtin/storage_framework_service_test.go]: It is fine to not assign to classicSlot AFAICS. In fact, you could probably remove it from StorageFrameworkServiceInterfaceSuite since you aren't doing anything with it. If it is decided that storage-framework-service would be provided outside of a snap, then we can start considering classic distro policy.

File: interfaces/builtin/storage_framework_service.go (hunk 8 above)
Comment (zyga) [inline interfaces/builtin/storage_framework_service.go]: This is done

File: interfaces/builtin/storage_framework_service.go (hunk 9 above)
Comment (zyga) [inline interfaces/builtin/storage_framework_service.go]: This is done

File: interfaces/builtin/storage_framework_service_test.go (hunk 14 above)
Comment (zyga) [inline interfaces/builtin/storage_framework_service_test.go]: I iterated on the unit tests. Please have a look if anything is missing.
